collector/metrics/metric_collector.go

```diff
@@ -0,0 +1,354 @@
+	}
+
+	for _, addr := range addrs {
+		data, err := getOneMetaServerMetrics(addr)
+		if err != nil {
+			log.Errorf(fmt.Sprintf("Get raw metrics from %s failed, err: %s", addr, err))
+			return
+		}
+		json_data := gjson.Parse(data)
+		for _, entity := range json_data.Array() {
+			for _, metric := range entity.Get("metrics").Array() {
+				var name string = metric.Get("name").String()
+				var mtype string = metric.Get("type").String()
+				var desc string = metric.Get("description").String()
+				log.Errorf("name:%s type:%s desc:%s", name, mtype, desc)
+				switch mtype {
+				case "counter":
+					if _, ok := counterMetricsMap_[name]; ok {
+						continue
+					}
+					counterMetric := promauto.NewCounterVec(prometheus.CounterOpts{
+						Name: name,
+						Help: desc,
+					}, []string{"level", "endpoint"})
+					counterMetricsMap_[name] = *counterMetric
+					break
+				case "gauge":
+					if _, ok := gaugeMetricsMap_[name]; ok {
+						continue
+					}
+					gaugeMetric := promauto.NewGaugeVec(prometheus.GaugeOpts{
+						Name: name,
+						Help: desc,
+					}, []string{"level", "endpoint"})
+					gaugeMetricsMap_[name] = *gaugeMetric
+					break
+				case "histogram":
+					if _, ok := histogramMetricsMap_[name]; ok {
+						continue
+					}
+					histogramMetric := promauto.NewHistogram(prometheus.HistogramOpts{
+						Name:    name,
+						Help:    desc,
+						Buckets: prometheus.LinearBuckets(-3, .1, 61),
+					})
+					histogramMetricsMap_[name] = histogramMetric
+					break
+				case "percentile":
+					if _, ok := summaryMetricsMap_[name]; ok {
+						continue
+					}
+					log.Errorf("name:%s", name)
+					summaryMetric := promauto.NewSummary(prometheus.SummaryOpts{
+						Name:    name,
+						Help:    desc,
+						Objectives: map[float64]float64{
+							0.5: 0.05, 0.9: 0.01, 0.95: 0.005, 0.99: 0.001, 0.999: 0.0001},
+					})
+					summaryMetricsMap_[name] = summaryMetric
+					break
+				default:
+					log.Errorf("Unsupport metric type %s", mtype)
+					break
+				}
+			}
+		}
+	}
+}
+
+func processAllMetaServerMetrics() {
+	var addrs []string
+	if dataSource_ == META_SERVER {
+		addrs = viper.GetStringSlice("meta_servers")
+	} else {
+		addrs = viper.GetStringSlice("replica_servers")
+	}
+	metric_by_table_id := make(map[string]Metrics, 128)
+	metric_of_cluster := make([]Metric, 128)
+	for _, addr := range addrs {
+		data, err := getOneMetaServerMetrics(addr)
+		if err != nil {
+			log.Errorf("failed to get data from %s, err %s", addr, err)
+			return
+		}
+		json_data := gjson.Parse(data)
+		for _, entity := range json_data.Array() {
+			etype := entity.Get("type").String()
+			switch etype {
+			case "replica":
+			case "partition":
+				table_id := entity.Get("attributes").Get("table_id").String()
+				mergeIntoTableLevel(entity.Get("metrics").Array(), table_id, &metric_by_table_id)
+				break
+			case "server":
+				mergeIntoClusterLevel(entity.Get("metrics").Array(), metric_of_cluster)
```

Comment:
The raw server level metrics are also needed to sink to Promethues, that is to say, some metrics are with attributes
```
level: server
endpoint: <host>
```